src/cpp/dds/domain/DomainParticipantImpl.cpp

```diff
@@ -0,0 +1,714 @@
+{
+    return rtps_participant_->getParticipantNames();
+}
+
+Subscriber* DomainParticipantImpl::create_subscriber(
+        const SubscriberQos& qos,
+        const fastrtps::SubscriberAttributes& att,
+        SubscriberListener* listen)
+{
+    if(att_.rtps.builtin.discovery_config.use_STATIC_EndpointDiscoveryProtocol)
+    {
+        if(att.getUserDefinedID() <= 0)
+        {
+            logError(PARTICIPANT,"Static EDP requires user defined Id");
+            return nullptr;
+        }
+    }
+
+    if(!att.unicastLocatorList.isValid())
+    {
+        logError(PARTICIPANT,"Unicast Locator List for Subscriber contains invalid Locator");
+        return nullptr;
+    }
+
+    if(!att.multicastLocatorList.isValid())
+    {
+        logError(PARTICIPANT," Multicast Locator List for Subscriber contains invalid Locator");
+        return nullptr;
+    }
+
+    if(!att.remoteLocatorList.isValid())
+    {
+        logError(PARTICIPANT,"Output Locator List for Subscriber contains invalid Locator");
+        return nullptr;
+    }
+
+    if(!att.qos.checkQos() || !att.topic.checkQos())
+    {
+        return nullptr;
+    }
+
+    SubscriberImpl* subimpl = new SubscriberImpl(this, qos, att, listen);
+    Subscriber* sub = new Subscriber(subimpl);
+    subimpl->user_subscriber_ = sub;
+    subimpl->rtps_participant_ = this->rtps_participant_;
+
+    // Create InstanceHandle for the new subscriber
+    GUID_t sub_guid = guid();
+    do
+    {
+        sub_guid.entityId = fastrtps::rtps::c_EntityId_Unknown;
+        rtps_participant_->get_new_entity_id(sub_guid.entityId);
+    } while (exists_entity_id(sub_guid.entityId));
+
+    InstanceHandle_t sub_handle(sub_guid);
+    subimpl->handle_ = sub_handle;
+
+    //SAVE THE PUBLISHER INTO MAPS
+    std::lock_guard<std::mutex> lock(mtx_subs_);
+    subscribers_by_handle_[sub_handle] = sub;
+    subscribers_[sub] = subimpl;
+
+    return sub;
+}
+
+
+const TypeSupport DomainParticipantImpl::find_type(
+        const std::string& type_name) const
+{
+    std::lock_guard<std::mutex> lock(mtx_types_);
+
+    auto type_it = types_.find(type_name);
+
+    if (type_it != types_.end())
+    {
+        return type_it->second;
+    }
+
+    return nullptr;
+}
+
+bool DomainParticipantImpl::register_type(
+        const TypeSupport type,
+        const std::string& type_name)
+{
+    TypeSupport t = find_type(type_name);
+
+    if (t != nullptr)
+    {
+        if (t == type)
+        {
+            return true;
+        }
+
+        logError(PARTICIPANT, "Another type with the same name '" << type_name << "' is already registered.");
+    }
+
+    if (type_name.size() <= 0)
+    {
+        logError(PARTICIPANT, "Registered Type must have a name");
+        return false;
+    }
+
+    logInfo(PARTICIPANT, "Type " << type_name << " registered.");
+    std::lock_guard<std::mutex> lock(mtx_types_);
+    types_.insert(std::make_pair(type_name, type));
+    return true;
```

Comment:
Add empty line above this line to more clarity

Reply:
This suggestion isn't in the coding style and it is a matter of taste. It could conflict with the "no empty lines except to separate blocks of functionality" rule.